src/elm/Persistence/InventoryDecoder.elm

```diff
@@ -0,0 +1,90 @@
+module Persistence.InventoryDecoder exposing (decodeInventory)
+
+import Data.GameConfiguration as GameConfiguration exposing (Level(..))
+import Data.Inventory as Inventory exposing (Inventory(..))
+import Json.Decode as Decode exposing (Decoder)
+import Json.Decode.Extra as Decode
+
+
+type alias MultiplierCounts =
+    { click : Int
+    , resources : List ( GameConfiguration.Level, Int )
+    }
+
+
+decodeInventory : Decode.Value -> Inventory
+decodeInventory value =
+    Decode.decodeValue inventoryDecoder value
+        |> Result.withDefault Inventory.initial
+
+
+inventoryDecoder : Decoder Inventory
+inventoryDecoder =
+    Decode.map3 Inventory.buildFromInitial
+        (Decode.at [ "value", "wallet" ] Decode.float)
+        (Decode.at [ "value", "resources" ] resourcesDecoder)
+        (Decode.at [ "value", "multipliers" ] multiplierCountsDecoder)
+
+
+decodeLevel : String -> Result String GameConfiguration.Level
+decodeLevel value =
+    case value of
+        "L1" ->
+            Ok L1
+
+        "L2" ->
+            Ok L2
+
+        "L3" ->
+            Ok L3
+
+        "L4" ->
+            Ok L4
+
+        "L5" ->
+            Ok L5
+
+        "L6" ->
+            Ok L6
+
+        "L7" ->
+            Ok L7
+
+        "L8" ->
+            Ok L8
+
+        _ ->
+            Err "Unable to decode level"
+
+
+decodeResourceCounts : List ( String, Int ) -> Decoder (List ( GameConfiguration.Level, Int ))
+decodeResourceCounts =
+    Decode.combine
+        << List.map
+            (\( levelString, value ) ->
+                case decodeLevel levelString of
+                    Ok level ->
+                        Decode.succeed ( level, value )
+
+                    Err error ->
+                        Decode.fail error
```

Comment:
This case looks like it's re-implementing `Json.Decode.Extra.fromResult` ?

Reply:
Not quite, since we are going from `Result String a -> Decoder (a, b)` and not `Result String a -> Decoder a`